### PR TITLE
port to support hdf5 1.12

### LINKED ISSUE #711: builds against hdf5 1.12.x break due to api changes in HDF5
Specifically  - the `H5Oget_info` compatibility macros changed. 

```
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:833:57: error: too few arguments to function ‘herr_t H5Oget_info3(hid_t, H5O_info2_t*, unsigned int)’
     herr_t h5_status = H5Oget_info(hdf5_id, &h5_obj_info);
                                                         ^
In file included from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Apublic.h:22:0,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/hdf5.h:23,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.hpp:17,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:14:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Opublic.h:187:15: note: declared here
 H5_DLL herr_t H5Oget_info3(hid_t loc_id, H5O_info2_t *oinfo, unsigned fields);
               ^~~~~~~~~~~~
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp: In function ‘bool conduit::relay::io::check_if_conduit_object_is_compatible_with_hdf5_tree(const conduit::Node&, const string&, hid_t, std::__cxx11::string&)’:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:956:57: error: too few arguments to function ‘herr_t H5Oget_info3(hid_t, H5O_info2_t*, unsigned int)’
     herr_t h5_status = H5Oget_info(hdf5_id, &h5_obj_info);
                                                         ^
In file included from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Apublic.h:22:0,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/hdf5.h:23,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.hpp:17,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:14:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Opublic.h:187:15: note: declared here
 H5_DLL herr_t H5Oget_info3(hid_t loc_id, H5O_info2_t *oinfo, unsigned fields);
               ^~~~~~~~~~~~
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp: In function ‘bool conduit::relay::io::check_if_conduit_list_is_compatible_with_hdf5_tree(const conduit::Node&, const string&, hid_t, std::__cxx11::string&)’:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:1024:57: error: too few arguments to function ‘herr_t H5Oget_info3(hid_t, H5O_info2_t*, unsigned int)’
     herr_t h5_status = H5Oget_info(hdf5_id, &h5_obj_info);
                                                         ^
In file included from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Apublic.h:22:0,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/hdf5.h:23,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.hpp:17,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:14:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Opublic.h:187:15: note: declared here
 H5_DLL herr_t H5Oget_info3(hid_t loc_id, H5O_info2_t *oinfo, unsigned fields);
               ^~~~~~~~~~~~
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp: In function ‘void conduit::relay::io::write_conduit_leaf_to_hdf5_group(const conduit::Node&, const string&, hid_t, const string&)’:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:1503:61: error: too few arguments to function ‘herr_t H5Oget_info_by_name3(hid_t, const char*, H5O_info2_t*, unsigned int, hid_t)’
                                                  H5P_DEFAULT);
                                                             ^
In file included from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Apublic.h:22:0,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/hdf5.h:23,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.hpp:17,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:14:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Opublic.h:188:15: note: declared here
 H5_DLL herr_t H5Oget_info_by_name3(hid_t loc_id, const char *name, H5O_info2_t *oinfo,
               ^~~~~~~~~~~~~~~~~~~~
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp: In function ‘void conduit::relay::io::write_conduit_empty_to_hdf5_group(hid_t, const string&, const string&)’:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:1568:61: error: too few arguments to function ‘herr_t H5Oget_info_by_name3(hid_t, const char*, H5O_info2_t*, unsigned int, hid_t)’
                                                  H5P_DEFAULT);
                                                             ^
In file included from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Apublic.h:22:0,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/hdf5.h:23,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.hpp:17,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:14:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Opublic.h:188:15: note: declared here
 H5_DLL herr_t H5Oget_info_by_name3(hid_t loc_id, const char *name, H5O_info2_t *oinfo,
               ^~~~~~~~~~~~~~~~~~~~
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp: In function ‘void conduit::relay::io::write_conduit_node_children_to_hdf5_group(const conduit::Node&, const string&, hid_t)’:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:1671:69: error: too few arguments to function ‘herr_t H5Oget_info_by_name3(hid_t, const char*, H5O_info2_t*, unsigned int, hid_t)’
                                                          H5P_DEFAULT);
                                                                     ^
In file included from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Apublic.h:22:0,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/hdf5.h:23,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.hpp:17,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:14:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Opublic.h:188:15: note: declared here
 H5_DLL herr_t H5Oget_info_by_name3(hid_t loc_id, const char *name, H5O_info2_t *oinfo,
               ^~~~~~~~~~~~~~~~~~~~
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp: In function ‘herr_t conduit::relay::io::h5l_iterate_traverse_op_func(hid_t, const char*, const H5L_info2_t*, void*)’:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:2001:48: error: too few arguments to function ‘herr_t H5Oget_info_by_name3(hid_t, const char*, H5O_info2_t*, unsigned int, hid_t)’
                                     H5P_DEFAULT);
                                                ^
In file included from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Apublic.h:22:0,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/hdf5.h:23,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.hpp:17,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:14:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Opublic.h:188:15: note: declared here
 H5_DLL herr_t H5Oget_info_by_name3(hid_t loc_id, const char *name, H5O_info2_t *oinfo,
               ^~~~~~~~~~~~~~~~~~~~
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:2029:53: error: ‘H5O_info2_t {aka struct H5O_info2_t}’ has no member named ‘addr’
             if ( h5_group_check (h5_od, h5_info_buf.addr) )
                                                     ^~~~
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp: In function ‘void conduit::relay::io::read_hdf5_group_into_conduit_node(hid_t, const string&, conduit::Node&)’:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:2119:48: error: too few arguments to function ‘herr_t H5Oget_info3(hid_t, H5O_info2_t*, unsigned int)’
                                    &h5_info_buf);
                                                ^
In file included from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Apublic.h:22:0,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/hdf5.h:23,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.hpp:17,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:14:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Opublic.h:187:15: note: declared here
 H5_DLL herr_t H5Oget_info3(hid_t loc_id, H5O_info2_t *oinfo, unsigned fields);
               ^~~~~~~~~~~~
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:2139:30: error: ‘H5O_info2_t {aka struct H5O_info2_t}’ has no member named ‘addr’
     h5_od.addr = h5_info_buf.addr;
                              ^~~~
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp: In function ‘void conduit::relay::io::read_hdf5_tree_into_conduit_node(hid_t, const string&, conduit::Node&)’:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:2360:49: error: too few arguments to function ‘herr_t H5Oget_info3(hid_t, H5O_info2_t*, unsigned int)’
     h5_status = H5Oget_info(hdf5_id,&h5_info_buf);
                                                 ^
In file included from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Apublic.h:22:0,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/hdf5.h:23,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.hpp:17,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:14:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Opublic.h:187:15: note: declared here
 H5_DLL herr_t H5Oget_info3(hid_t loc_id, H5O_info2_t *oinfo, unsigned fields);
               ^~~~~~~~~~~~
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp: In function ‘void conduit::relay::io::hdf5_group_list_child_names(hid_t, const string&, std::vector<std::__cxx11::basic_string<char> >&)’:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:3024:55: error: too few arguments to function ‘herr_t H5Oget_info_by_name3(hid_t, const char*, H5O_info2_t*, unsigned int, hid_t)’
                                            H5P_DEFAULT);
                                                       ^
In file included from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Apublic.h:22:0,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/hdf5.h:23,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.hpp:17,
                 from /masonry/build-mb-develop-ci-smoke/thirdparty_shared/conduit-v0.7.1/src/libs/relay/conduit_relay_io_hdf5.cpp:14:
/masonry/build-mb-develop-ci-smoke/thirdparty_shared/third_party/hdf5/1.12.0/ci/include/H5Opublic.h:188:15: note: declared here
 H5_DLL herr_t H5Oget_info_by_name3(hid_t loc_id, const char *name, H5O_info2_t *oinfo,
               ^~~~~~~~~~~~~~~~~~~~
```

### PR DESCRIPTION
**Pull request by cyrush**
resolves #711

### CODE REVIEW

**Comment by jandrej** [comment]
ping @cyrush 

I'm using the patch set during development of a code where conduit is a dependency and didn't run into any issues.

**Comment by cyrush** [comment]
@jandrej thanks for the report -- however I will share -- I know there are outstanding issues.
Depending on your use cases you may not hit them - but the are related to reading HDF5 trees.

File: src/libs/relay/conduit_relay_io_hdf5.cpp [hunk 1]
```diff
@@ -2369,8 +2469,14 @@ read_hdf5_group_into_conduit_node(hid_t hdf5_group_id,
 {
     // get info, we need to get the obj addr for cycle tracking
     H5O_info_t h5_info_buf;
+#if H5_VERSION_GE(1, 12, 0)
+    herr_t h5_status = H5Oget_info(hdf5_group_id,
+                                   &h5_info_buf,
+                                   H5P_DEFAULT);
```
Comment (kennyweiss) [inline src/libs/relay/conduit_relay_io_hdf5.cpp]:
```suggestion
                                   H5O_INFO_ALL;
```

File: src/libs/relay/conduit_relay_io_hdf5.cpp [hunk 2]
```diff
@@ -2681,10 +2791,14 @@ read_hdf5_tree_into_conduit_node(hid_t hdf5_id,
                                  const Node &opts,
                                  Node &dest)
 {
-    herr_t     h5_status = 0;
     H5O_info_t h5_info_buf;
 
-    h5_status = H5Oget_info(hdf5_id,&h5_info_buf);
+#if H5_VERSION_GE(1, 12, 0)
+    herr_t h5_status = H5Oget_info(hdf5_id,&h5_info_buf,H5P_DEFAULT);
```
Comment (kennyweiss) [inline src/libs/relay/conduit_relay_io_hdf5.cpp]:
```suggestion
    herr_t h5_status = H5Oget_info(hdf5_id,&h5_info_buf,H5O_INFO_ALL);
```

**Review by kennyweiss — APPROVED** [review]
Thanks @cyrush -- this came in really handy in https://github.com/LLNL/axom/pull/597 since the newer `vcpkg` is using `hdf5@1.12`.  

I created a patch file in axom based on these changes (https://github.com/LLNL/axom/pull/597/files#diff-3c429375dab810c66dbf34487bd62c039e085780d07640b0c1aa2c413766a2aa), which I'll be happy to remove once this PR is merged. 
I pointed out two proposed changes based on some failing Axom/sidre tests (see: https://github.com/LLNL/axom/pull/597/commits/40c07404b83b5d1ddf974990bb4cd1612afee091)

**Comment by cyrush** [comment]
thanks for looking at this @kennyweiss  -- I'll try these changes!

**Comment by cyrush** [comment]
@kevinhkhuynh  -- Q for you: For the magic that we use to upgrade a fixed dataset to a extensible dataset.

I am wondering your thoughts on changing:

`hdf5_write(const Node &node, hid_t hdf5_id, const Node &opts);`

To take a reference to the id, that way if we close and change it everything stays consistent for the caller:

`hdf5_write(const Node &node, hid_t &hdf5_id, const Node &opts);`

what do you think?

From the calling cases in our tests this seems like a good evolution.

**Comment by cyrush** [comment]
@kennyweiss  @jandrej 

Good news -- I tracked down the unit test failures. They were limited to a small set of the features. Turns out they occurred in hdf 1.10.x and those fixes apply here as well.

**Comment by kevinhkhuynh** [comment]
> @kevinhkhuynh -- Q for you: For the magic that we use to upgrade a fixed dataset to a extensible dataset.
> 
> I am wondering your thoughts on changing:
> 
> `hdf5_write(const Node &node, hid_t hdf5_id, const Node &opts);`
> 
> To take a reference to the id, that way if we close and change it everything stays consistent for the caller:
> 
> `hdf5_write(const Node &node, hid_t &hdf5_id, const Node &opts);`
> 
> what do you think?
> 
> From the calling cases in our tests this seems like a good evolution.

@cyrush  This makes sense to me. I agree with this change.

**Comment by cyrush** [comment]
@kevinhkhuynh thanks!

**Pull request merged**